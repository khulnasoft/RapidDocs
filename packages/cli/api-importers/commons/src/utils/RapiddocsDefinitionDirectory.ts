import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { RelativeFilePath, sep } from "@khulnasoft/path-utils";

export class RapiddocsDefinitionDirectory {
    private files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};
    private directories: Record<string, RapiddocsDefinitionDirectory> = {};

    public getAllFiles(): Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> {
        const files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};

        const walk = (root: RapiddocsDefinitionDirectory, currentPath?: string) => {
            for (const [relativeFilePath, definition] of Object.entries(root.files)) {
                const fullRelativeFilePath =
                    currentPath != null
                        ? RelativeFilePath.of(`${currentPath}${sep}${relativeFilePath}`)
                        : RelativeFilePath.of(relativeFilePath);
                files[fullRelativeFilePath] = definition;
            }
            const sortedDirectories = Object.keys(root.directories).sort();
            for (const directory of sortedDirectories) {
                const nextPath = currentPath != null ? `${currentPath}${sep}${directory}` : directory;
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const nextDirectory = root.directories[directory]!;
                walk(nextDirectory, nextPath);
            }
        };

        walk(this);

        return files;
    }

    public getOrCreateFile(relativeFilePath: RelativeFilePath): RawSchemas.DefinitionFileSchema {
        return this.getOrCreateFileRecursive(relativeFilePath.split(sep));
    }

    private getOrCreateFileRecursive(pathParts: string[]): RawSchemas.DefinitionFileSchema {
        if (pathParts.length === 1) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            return (this.files[RelativeFilePath.of(pathParts[0]!)] ??= {});
        }
        const [directory, ...remainingPath] = pathParts;
        if (directory == null) {
            throw new Error(`Internal error; cannot add file with path: ${pathParts}`);
        }
        if (!this.directories[directory]) {
            this.directories[directory] = new RapiddocsDefinitionDirectory();
        }
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return this.directories[directory]!.getOrCreateFileRecursive(remainingPath);
    }
}
