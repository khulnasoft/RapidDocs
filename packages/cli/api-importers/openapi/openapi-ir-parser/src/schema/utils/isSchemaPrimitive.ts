import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaPrimitive(schema: Schema): boolean {
    if (schema.type === "primitive") {
        return true;
    } else if (schema.type === "nullable" || schema.type === "optional") {
        return isSchemaPrimitive(schema.value);
    }
    return false;
}
