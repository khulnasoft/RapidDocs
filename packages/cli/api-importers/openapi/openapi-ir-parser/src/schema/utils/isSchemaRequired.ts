import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaRequired(schema: Schema): boolean {
    return schema.type !== "optional" && schema.type !== "nullable";
}
