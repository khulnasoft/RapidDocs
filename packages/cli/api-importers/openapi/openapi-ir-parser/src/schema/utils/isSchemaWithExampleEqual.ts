import { SchemaWithExample, isSchemaEqual } from "@khulnasoft/openapi-ir";

import { convertSchemaWithExampleToSchema } from "./convertSchemaWithExampleToSchema";

// only diffs the schema, not the example
export function isSchemaWithExampleEqual(a: SchemaWithExample, b: SchemaWithExample): boolean {
    return isSchemaEqual(convertSchemaWithExampleToSchema(a), convertSchemaWithExampleToSchema(b));
}
