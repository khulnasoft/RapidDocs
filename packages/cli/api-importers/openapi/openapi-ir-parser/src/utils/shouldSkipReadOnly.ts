import { RapiddocsOpenapiIr } from "@khulnasoft/openapi-ir";

export function shouldSkipReadOnly(method: RapiddocsOpenapiIr.HttpMethod): boolean | undefined {
    return method === "POST" || method === "PUT" || method === "PATCH";
}
