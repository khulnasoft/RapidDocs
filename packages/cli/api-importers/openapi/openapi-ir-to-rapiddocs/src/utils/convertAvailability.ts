import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Availability } from "@khulnasoft/openapi-ir";

export function convertAvailability(
    availability: Availability | undefined
): RawSchemas.AvailabilityUnionSchema | undefined {
    switch (availability) {
        case Availability.Deprecated:
            return "deprecated";
        case Availability.Beta:
            return "pre-release";
        case Availability.GenerallyAvailable:
            return "generally-available";
        default:
            return undefined;
    }
}
