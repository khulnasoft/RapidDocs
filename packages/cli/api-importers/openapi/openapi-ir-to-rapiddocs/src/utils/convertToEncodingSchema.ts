import { assertNever } from "@khulnasoft/core-utils";
import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Encoding } from "@khulnasoft/openapi-ir";

export function convertToEncodingSchema(encoding: Encoding): RawSchemas.EncodingSchema {
    switch (encoding.type) {
        case "protobuf":
            return {
                proto: {
                    type: encoding.typeName
                }
            };
        default:
            assertNever(encoding.type);
    }
}
