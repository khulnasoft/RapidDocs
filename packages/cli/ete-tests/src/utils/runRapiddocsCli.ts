import { Options } from "execa";
import path from "path";

import { loggingExeca } from "@khulnasoft/logging-execa";

export async function runRapiddocsCli(args: string[], options?: Options): Promise<loggingExeca.ReturnValue> {
    return loggingExeca(undefined, "node", [path.join(__dirname, "../../../cli/dist/dev/cli.cjs"), ...args], {
        ...options,
        env: {
            ...options?.env,
            RAPIDDOCS_TOKEN: process.env.RAPIDDOCS_ORG_TOKEN_DEV
        },
        doNotPipeOutput: options?.reject === false
    });
}
