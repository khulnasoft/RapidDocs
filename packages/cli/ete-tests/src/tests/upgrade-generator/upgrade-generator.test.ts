/* eslint-disable jest/no-disabled-tests */
import { cp, readFile } from "fs/promises";
import path from "path";
import tmp from "tmp-promise";

import { AbsoluteFilePath, RelativeFilePath, getDirectoryContents, join } from "@khulnasoft/fs-utils";

import { runRapiddocsCli } from "../../utils/runRapiddocsCli";

const FIXTURES_DIR = path.join(__dirname, "fixtures");

describe("rapiddocs generator upgrade", () => {
    it("rapiddocs generator upgrade", async () => {
        // Create tmpdir and copy contents
        const tmpDir = await tmp.dir();
        const directory = AbsoluteFilePath.of(tmpDir.path);

        await cp(FIXTURES_DIR, directory, { recursive: true });

        await runRapiddocsCli(["generator", "upgrade"], {
            cwd: directory
        });

        const outputFile = join(directory, RelativeFilePath.of("version.txt"));
        await runRapiddocsCli(
            [
                "generator",
                "get",
                "--group",
                "python-sdk",
                "--generator",
                "rapiddocsapi/rapiddocs-python-sdk",
                "--version",
                "-o",
                outputFile
            ],
            {
                cwd: directory
            }
        );

        expect(JSON.parse((await readFile(outputFile)).toString()).version).not.toEqual("3.0.0");
    }, 60_000);

    it("rapiddocs generator upgrade with filters", async () => {
        // Create tmpdir and copy contents
        const tmpDir = await tmp.dir();
        const directory = AbsoluteFilePath.of(tmpDir.path);

        await cp(FIXTURES_DIR, directory, { recursive: true });

        await runRapiddocsCli(
            [
                "generator",
                "upgrade",
                "--group",
                "python-sdk",
                "--generator",
                "rapiddocsapi/rapiddocs-python-sdk",
                "--include-major"
            ],
            {
                cwd: directory
            }
        );

        const outputFile = join(directory, RelativeFilePath.of("version.txt"));
        await runRapiddocsCli(
            [
                "generator",
                "get",
                "--group",
                "python-sdk",
                "--generator",
                "rapiddocsapi/rapiddocs-python-sdk",
                "--version",
                "-o",
                outputFile
            ],
            {
                cwd: directory
            }
        );

        expect(JSON.parse((await readFile(outputFile)).toString()).version).not.toEqual("3.0.0");
    }, 60_000);

    it("rapiddocs generator help commands", async () => {
        // Create tmpdir and copy contents
        const tmpDir = await tmp.dir();
        const directory = AbsoluteFilePath.of(tmpDir.path);

        await cp(FIXTURES_DIR, directory, { recursive: true });

        expect(
            (
                await runRapiddocsCli(["generator", "--help"], {
                    cwd: directory,
                    reject: false
                })
            ).stdout
        ).toMatchSnapshot();

        expect(
            (
                await runRapiddocsCli(["generator", "upgrade", "--help"], {
                    cwd: directory,
                    reject: false
                })
            ).stdout
        ).toMatchSnapshot();
    }, 60_000);

    it("rapiddocs generator upgrade majors", async () => {
        // Create tmpdir and copy contents
        const tmpDir = await tmp.dir();
        const directory = AbsoluteFilePath.of(tmpDir.path);

        await cp(FIXTURES_DIR, directory, { recursive: true });

        await runRapiddocsCli(
            ["generator", "upgrade", "--group", "shouldnt-upgrade", "--generator", "rapiddocsapi/rapiddocs-python-sdk"],
            {
                cwd: directory
            }
        );

        const outputFile = join(directory, RelativeFilePath.of("version.txt"));
        await runRapiddocsCli(
            [
                "generator",
                "get",
                "--group",
                "shouldnt-upgrade",
                "--generator",
                "rapiddocsapi/rapiddocs-python-sdk",
                "--version",
                "-o",
                outputFile
            ],
            {
                cwd: directory
            }
        );

        expect(JSON.parse((await readFile(outputFile)).toString()).version).toEqual("2.16.0");

        await runRapiddocsCli(
            [
                "generator",
                "upgrade",
                "--group",
                "python-sdk",
                "--generator",
                "rapiddocsapi/rapiddocs-python-sdk",
                "--include-major"
            ],
            {
                cwd: directory
            }
        );

        const outputFileNewMajor = join(directory, RelativeFilePath.of("version-new.txt"));
        await runRapiddocsCli(
            [
                "generator",
                "get",
                "--group",
                "python-sdk",
                "--generator",
                "rapiddocsapi/rapiddocs-python-sdk",
                "--version",
                "-o",
                outputFileNewMajor
            ],
            {
                cwd: directory
            }
        );

        expect(JSON.parse((await readFile(outputFileNewMajor)).toString()).version).not.toEqual("2.16.0");
    }, 60_000);

    it("rapiddocs generator upgrade list", async () => {
        // Create tmpdir and copy contents
        const tmpDir = await tmp.dir();
        const directory = AbsoluteFilePath.of(tmpDir.path);

        await cp(FIXTURES_DIR, directory, { recursive: true });

        expect(
            (
                await runRapiddocsCli(
                    [
                        "generator",
                        "upgrade",
                        "--group",
                        "python-sdk",
                        "--generator",
                        "rapiddocsapi/rapiddocs-python-sdk",
                        "--include-major",
                        "--list"
                    ],
                    {
                        cwd: directory
                    }
                )
            ).stdout
        ).toMatchSnapshot();
    }, 60_000);

    it.skip("rapiddocs generator upgrade message", async () => {
        const tmpDir = await tmp.dir();
        const directory = AbsoluteFilePath.of(tmpDir.path);

        await cp(FIXTURES_DIR, directory, { recursive: true });

        const outputFileNewMajor = join(directory, RelativeFilePath.of("version-new.txt"));

        expect(
            (
                await runRapiddocsCli(
                    [
                        "generator",
                        "get",
                        "--group",
                        "python-sdk",
                        "--generator",
                        "rapiddocsapi/rapiddocs-python-sdk",
                        "--version",
                        "-o",
                        outputFileNewMajor
                    ],
                    {
                        cwd: directory
                    }
                )
            ).stdout
        ).toMatchSnapshot();
    }, 60_000);
});
