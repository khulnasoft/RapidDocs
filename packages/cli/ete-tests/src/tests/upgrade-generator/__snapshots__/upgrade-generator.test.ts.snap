// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rapiddocs generator upgrade > rapiddocs generator help commands 1`] = `
"rapiddocs-dev generator

Operate on the generators within your Rapiddocs configuration

Commands:
  rapiddocs-dev generator upgrade  Upgrades the specified generator in generators.yml
                              to the latest stable version.

Options:
  --help       Show help                                               [boolean]
  --log-level
          [choices: "trace", "debug", "info", "warn", "error"] [default: "info"]"
`;

exports[`rapiddocs generator upgrade > rapiddocs generator help commands 2`] = `
"rapiddocs-dev generator upgrade

Upgrades the specified generator in generators.yml to the latest stable version.

Options:
  --help           Show help                                           [boolean]
  --log-level
          [choices: "trace", "debug", "info", "warn", "error"] [default: "info"]
  --generator      The type of generator to upgrade, ex:
                   \`rapiddocs-typescript-node-sdk\`.                          [string]
  --group          The group in which the generator is located, if group is not
                   specified, the all generators of the specified type will be
                   upgraded.                                            [string]
  --api            The API to upgrade the generator for. If not specified, the
                   generator will be upgraded for all APIs.             [string]
  --include-major  Whether or not to include major versions within the upgrade.
                   Defaults to false, meaning major versions will be skipped.
                                                      [boolean] [default: false]
  --channel                                                [choices: "GA", "RC"]
  --list           When specified, a list of available upgrades will be
                   displayed, but no upgrade will be taken.
                                                      [boolean] [default: false]"
`;

exports[`rapiddocs generator upgrade > rapiddocs generator upgrade list 1`] = `
"Warnings for generators.yml:
	Using "api.path" is deprecated. Please use "api.specs[].openapi" or "api.specs[].asyncapi" instead.
[api]: Warnings for generators.yml:
[api]: 	Using "api.path" is deprecated. Please use "api.specs[].openapi" or "api.specs[].asyncapi" instead.
              ╭──────────────────────────────────────────────────╮
              │                                                  │
              │                Generator Upgrades                │
              │                                                  │
              │   Python SDK (Group: python-sdk) 3.0.0 → 4.2.7   │
              │                                                  │
              │                                                  │
              ╰──────────────────────────────────────────────────╯"
`;

exports[`rapiddocs generator upgrade > rapiddocs generator upgrade message 1`] = `
"         ╭────────────────────────────────────────────────────────────╮
         │                                                            │
         │                     Upgrades available                     │
         │                                                            │
         │                                                            │
         │           Java SDK (Group: local) 0.0.1 → 0.10.1           │
         │       Python SDK (Group: python-sdk) 3.0.0 → 3.10.8        │
         │                                                            │
         │   Run rapiddocs generator upgrade to upgrade your generators.   │
         │                                                            │
         ╰────────────────────────────────────────────────────────────╯
"
`;
